File: java/ValidSudoku.java
/*
 * Filename: ValidSudoku.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Sun 22 Mar 2015 04:01:16 PM EDT
 */

public class Solution {
	public boolean isValidSudoku(char[][] board) {
		int len = board.length;
		for (int i = 0; i < len; i++) {
			for (int j = 0; j < len; j++) {
				if (board[i][j] <= '9' && board[i][j] >= '1') {
					for (int k = 0; k < len; k++) {
						if (k != j && board[i][k] == board[i][j]) {
							return false;
						}

						if (k != i && board[k][j] == board[i][j]) {
							return false;
						}

						int iBlock = i / 3;
						int jBlock = j / 3;

						for (int m = 3 * iBlock; m < 3 * (iBlock + 1); m++) {
							for (int n = 3 * jBlock; n < 3 * (jBlock + 1); n++) {
								if (m != i && n != j && board[m][n] == board[i][j]) {
									return false;
								}
							}
						}
					}
				}
			}
		}

		return true;
	}
}

File: cpp/valid_sudoku.cpp
/*
 * Filename: valid_sudoku.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Sun 01 Nov 2015 10:24:08 PM EST
 * Description: 
 */

#include <iostream>
#include <map>
#include <vector>

using namespace std;

class Solution {
	public:
		bool isValidSudoku(vector<vector<char> > &board)
		{
			map<char, bool> row;
			map<char, bool> col;
			map<char, bool> block;

			for (size_t i = 0; i < 9; i++)
			{
				col.clear();
				row.clear();

				for (size_t j = 0; j < 9; j++)
				{
					if (board[i][j] != '.')
					{
						if (col[board[i][j]])
						{
							return false;
						}
						else
						{
							col[board[i][j]] = true;
						} 
					}

					if (board[j][i] != '.')
					{
						if (row[board[j][i]])
						{
							return false;
						}
						else
						{
							row[board[j][i]] = true;
						} 
					}
				}
			}

			for (size_t ii = 0; ii < 9; ii += 3)
			{
				for (size_t jj = 0; jj < 9; jj += 3)
				{
					block.clear();

					for (size_t i = ii; i < ii+3; i++)
					{
						for (size_t j = jj; j < jj+3; j++)
						{
							if (board[i][j] != '.')
							{
								if (block[board[i][j]])
								{
									return false;
								}
								else
								{
									block[board[i][j]] = true;
								} 
							}
						}
					}        
				}
			}

			return true;
		}
};
